pkg/skaffold/build/kaniko/run.go

```diff
@@ -164,3 +198,27 @@ func gcsDelete(ctx context.Context, bucket, path string) error {
 
 	return c.Bucket(bucket).Object(path).Delete(ctx)
 }
+
+func configMapCreate(artifact *v1alpha3.Artifact, ns, tag string) error {
+	logrus.Infof("Creating config map %s", configMapName(tag))
+	paths, err := docker.GetDependencies(artifact.Workspace, artifact.DockerArtifact)
+	if err != nil {
+		return errors.Wrap(err, "getting sources for configmap")
+	}
+	var files []string
+	for _, path := range paths {
+		files = append(files, []string{"--from-file", path}...)
+	}
+	cmd := exec.Command("kubectl", append([]string{"create", "configmap", configMapName(tag), "-n", ns}, files...)...)
```

Comment:
I'm not super familiar with this strategy, but this just stores the actual files in configmaps and mounts them? AFAIK the size limit for configmaps is the size limit for keys in etcd, which is 1.5 MB by default.

Have you tried this with larger repositories?

Reply:
I looked into it and you're right, this probably won't work with larger repositories.

What if we created a tarball of the files and mounted that in via configMap? Would require adding a `--context=tar://` option to kaniko but that should be pretty easy.

Reply:
I think it's the overall size

Reply:
What if we use Persitent Volumes? It would be a bit of a pain to maintain them - we'd have to take care of creating and cleaning up the PVs and figure out how to interact across namespaces as well potentially, but it would free the users from having to depend on any external storage solutions. If we figure this out, this might be useful for BuildCRD too.

Reply:
Another idea: we could enable this feature with ConfigMaps for smaller repos (<1.5MB) - if the repo is too big, it won't work. This would enable most of the users. In the meantime we can design a PV based approach for larger use cases.

Reply:
Using etcd for build context storage is going to make all of the apiserver interactions much slower and more intensive. 

ReadWrite PVCs are the right abstraction for context storage that we are also going to upload to.

Reply:
> "Using etcd for build context storage is going to make all of the apiserver interactions much slower and more intensive." 

why is that? for a single <1.5MB key in etcd every once in a while on a dev cluster?

Reply:
WDYT of using an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) to hold the build context? We could have an init container that runs while we `kubectl cp` the buildcontext into the emptyDir volume, and then mount that into the kaniko container.

I think the emptyDir can use up to however much memory is allocated to the cluster. 

It would allow for easier cleanup than a PV/PVC, although would be slower since we have to copy the entire buildcontext over for every dev loop.

Reply:
Thats assuming that the entire configmap fits in a 1.5 MB key.

If it doesn't:

- If you don't shard, then a very small number of repos are going to be able to fit in 1.5 MB.
- If you shard, you end up with a key per 1.5 MB, which for the skaffold repo would be 110 keys (164M/1.5M). The storage limit for etcd v3 is 2 GB but supports up to 8 GB. I'm not sure what the performance would be on that, but using 8% of your etcd for a build context could easily reach a performance limit pretty quickly on a shared cluster.

Reply:
I like @priyawadhwa's idea

* use the skaffold docker context generation to create a tarball as usual
* run the kaniko pod with an init container that waits for a special file to be written to signify completion
* kubectl cp the tarball over to the emptyDir
* kubectl cp the special completion file
* init container completes
* kaniko runs

On the other hand, if we reuse a PV to store the build context, we can do incremental build context uploads, which will be very fast.

Reply:
yeah, if 1.5MB is not good for most of the repos, then configmap doesn't work. 
+100 on @priyawadhwa's idea - and switch over to PVs as a follow-up feature for incremental builds later down the line if there's enough usage to warrant it